### PR TITLE
Driver for the Wolfson WM8904 audio codec

### PR DESCRIPTION
**Pull request by VitekST**
This PR brings a driver for the Wolfson/Cirrus WM8904 audio codec, based on original work of @hakehuang (recently closed PR: https://github.com/zephyrproject-rtos/zephyr/pull/66609#event-13467395343). The driver was developed and tested on the `mimxrt595_evk` and `mimxrt685_evk` boards. Following improvements were made to the driver itself:
- decoupled from `hal_nxp`
- input routing (multiplexing)
- input mute and volume control
- output mute and volume control
- omission of dead code

Following modifications were needed to implement it:
- modified `clock_control_mcux_syscon.c` to provide frequency for the MCLK signal
- extended the `codec` API to handle input and output routing, extra sample rates and configuration for stream directions (playback, capture, duplex)

### CODE REVIEW

**Comment by marc-hb** [comment]
I added @dbaluta and @iuliana-prodan as reviewers. I didn't realize this went over the 15 reviewers limit. When this happens, Github normally warns about this limit and blocks the addition. But that time Github dropped random other reviewers instead! Temporary glitch? You have been warned.

I tried to cleanup that mess, hopefully the list is back to where it was.

File: samples/drivers/i2s/i2s_codec/sample.yaml [hunk 1]
```diff
@@ -0,0 +1,13 @@
+sample:
+  name: codec sample
+tests:
+  sample.drivers.i2s.codec:
+    tags: i2s
+    filter: dt_alias_exists("i2s_codec")
```
Comment (hakehuang) [inline samples/drivers/i2s/i2s_codec/sample.yaml:6]: this filter will make the ci skip this case, and as you have add the platform_allow, either add this alias to dts or remove it.

Comment (VitekST) [reply samples/drivers/i2s/i2s_codec/sample.yaml:6]: Will remove then.

File: include/zephyr/audio/codec.h [hunk 2]
```diff
@@ -34,56 +34,66 @@ extern "C" {
...
+	AUDIO_PCM_RATE_11K = 11025,   /**< 11.025 kHz sample rate*/
+	AUDIO_PCM_RATE_16K = 16000,   /**< 16 kHz sample rate */
+	AUDIO_PCM_RATE_22K = 22050,   /**< 22.05 kHz sample rate*/
+	AUDIO_PCM_RATE_24K = 24000,   /**< 24 kHz sample rate */
+	AUDIO_PCM_RATE_32K = 32000,   /**< 32 kHz sample rate */
+	AUDIO_PCM_RATE_44P1K = 44100, /**< 44.1 kHz sample rate */
+	AUDIO_PCM_RATE_48K = 48000,   /**< 48 kHz sample rate */
+	AUDIO_PCM_RATE_96K = 96000,   /**< 96 kHz sample rate */
+	AUDIO_PCM_RATE_192K = 192000, /**< 192 kHz sample rate */
 } audio_pcm_rate_t;
```
Comment (dbaluta) [inline include/zephyr/audio/codec.h:48]: I don't undertand this change? You seem to add some tabs / space. This seems not to be related with the subject of the patch.  api: codec: Extend API 

If you want to do some cleanups then it should be a separate patch. Otherwise, this doesn't bring any value as it is in this patch.

One patch should mostly do one thing!

File: drivers/audio/Kconfig.wm8904 [hunk 3]
```diff
@@ -0,0 +1,10 @@
+# Copyright (c) 2024, NXP
+# SPDX-License-Identifier: Apache-2.0
```
Comment (dbaluta) [inline drivers/audio/Kconfig.wm8904:2]: Official way NXP adds copyright is. `Copyright 2024 NXP`.

File: include/zephyr/audio/codec.h (hunk 2 above)
Comment (VitekST) [reply include/zephyr/audio/codec.h:48]: I just invoked `clang-format` as recommended by contribution guidelines (https://docs.zephyrproject.org/latest/contribute/guidelines.html#clang-format) to take care of code formatting. Is this something that shouldn't be done when changing existing files?

Comment (dbaluta) [reply include/zephyr/audio/codec.h:48]: > I just invoked `clang-format` as recommended by contribution guidelines (https://docs.zephyrproject.org/latest/contribute/guidelines.html#clang-format) to take care of code formatting. Is this something that shouldn't be done when changing existing files?

yes, this shouldn't be taken into consideration for existing code. Only for new code added. If you still want to make the change you can make it as a separate PR. 

But maybe that is a task after you get this merged.

File: soc/nxp/imxrt/imxrt5xx/cm33/soc.c [hunk 4]
```diff
@@ -85,6 +85,7 @@ const clock_frg_clk_config_t g_frg12Config_clock_init = {
 #define BOARD_USB_PHY_TXCAL45DM (0x06U)
 #endif
 
+
 /* System clock frequency. */
```
Comment (dbaluta) [inline soc/nxp/imxrt/imxrt5xx/cm33/soc.c:89]: There is no need for a new line here.

**Comment by hakehuang** [comment]
CI board testing PASS.

**Review by hakehuang — DISMISSED** [review]

**Review by dbaluta — DISMISSED** [review]

File: include/zephyr/audio/codec.h [hunk 5]
```diff
@@ -35,7 +35,9 @@ extern "C" {
  */
 typedef enum {
 	AUDIO_PCM_RATE_8K	= 8000,		/**< 8 kHz sample rate */
+	AUDIO_PCM_RATE_11K	= 11025,	/**< 11.025 kHz sample rate */
 	AUDIO_PCM_RATE_16K	= 16000,	/**< 16 kHz sample rate */
+	AUDIO_PCM_RATE_22K	= 22050,	/**< 24 kHz sample rate */
```
Comment (hongshui3000) [inline include/zephyr/audio/codec.h:40]: The comment here should be 22.050kHz instead of 24 kHz

File: samples/drivers/i2s/i2s_codec/Kconfig [hunk 6]
```diff
@@ -0,0 +1,29 @@
...
+	int "number of i2s transfer buffer"
+	default 2
+	help
+	  i2s init buffers used in transfer
+
+config SAMPLE_FREQ
+	int "sample rate"
+	default 48000
+	help
+	  sample rate
```
Comment (decsny) [inline samples/drivers/i2s/i2s_codec/Kconfig:16]: these help texts are not sufficient, remember this will show in the menuconfig ui for users, it should have proper capitalization and grammar too

File: soc/nxp/imxrt/imxrt6xx/cm33/soc.c
Comment (decsny) [inline soc/nxp/imxrt/imxrt6xx/cm33/soc.c:1]: you can squash the soc commit for the 685 to the soc commit for the 595, same with the board commit to the other board commit, since they are small and doing the same thing, it makes the git history more consumable and less commit objects in the metadata

**Review by decsny — DISMISSED** [review]

File: samples/drivers/i2s/i2s_codec/Kconfig [hunk 7]
```diff
@@ -0,0 +1,31 @@
...
+	help
+	  Controls the initial count of audio data blocks, which are (optionally)
+	  filled by data from the DMIC peripheral and played back by the I2S
+	  output perihperal.
+
+config SAMPLE_FREQ
+	int "Sample rate"
+	default 48000
+	help
+	  Sample frequency of the system.
```
Comment (decsny) [inline samples/drivers/i2s/i2s_codec/Kconfig:18]: please be more specific. what is the "system"?

File: samples/drivers/i2s/i2s_codec/Kconfig [hunk 8]
```diff
@@ -0,0 +1,31 @@
...
+
+config I2S_INIT_BUFFERS
+	int "Initial count of audio data blocks"
+	default 2
+	help
+	  Controls the initial count of audio data blocks, which are (optionally)
+	  filled by data from the DMIC peripheral and played back by the I2S
+	  output perihperal.
+
+config SAMPLE_FREQ
```
Comment (decsny) [inline samples/drivers/i2s/i2s_codec/Kconfig:14]: namespace as I2S_SAMPLE_FREQ?

File: samples/drivers/i2s/i2s_codec/Kconfig (hunk 7 above)
Comment (VitekST) [reply samples/drivers/i2s/i2s_codec/Kconfig:18]: Well, the microcontroller along with the codec. The sample frequency is the only one configured. The codec is configured for that rate and samples are processed at that rate. Can't think of a better designation, honestly.

File: samples/drivers/i2s/i2s_codec/Kconfig (hunk 8 above)
Comment (VitekST) [reply samples/drivers/i2s/i2s_codec/Kconfig:14]: I wouldn't do that, is this frequency is used also for the DMIC peripheral.

File: samples/drivers/i2s/i2s_codec/boards/mimxrt595_evk_mimxrt595s_cm33.overlay [hunk 9]
```diff
@@ -0,0 +1,24 @@
+/ {
+	aliases {
+		i2s-codec-rx = &i2s0;
+		i2s-codec-tx = &i2s1;
+	};
+};
+
+&i3c0 {
+	clk-divider = <20>;
+
```
Comment (butok) [inline samples/drivers/i2s/i2s_codec/boards/mimxrt595_evk_mimxrt595s_cm33.overlay:10]: blank line

Comment (VitekST) [reply samples/drivers/i2s/i2s_codec/boards/mimxrt595_evk_mimxrt595s_cm33.overlay:10]: @butok Intentional - I like to semantically separate blocks of code.

File: dts/bindings/audio/wolfson,wm8904.yaml [hunk 10]
```diff
@@ -0,0 +1,25 @@
+# Copyright (c) 2023, NXP
+# SPDX-License-Identifier: Apache-2.0
+
```
Comment (dbaluta) [inline dts/bindings/audio/wolfson,wm8904.yaml:3]: same here: Use `Copyright 2024 NXP`

File: drivers/audio/wm8904.h [hunk 11]
```diff
@@ -0,0 +1,116 @@
+/*
+ * Copyright (c) 2023 NXP Semiconductor INC.
+ * All rights reserved.
+ *
```
Comment (dbaluta) [inline drivers/audio/wm8904.h:4]: Same here. Fix the copyright.

File: drivers/audio/wm8904.c [hunk 12]
```diff
@@ -0,0 +1,760 @@
+/*
+ * Copyright (c) 2023 NXP Semiconductor INC.
+ * All rights reserved.
+ *
```
Comment (dbaluta) [inline drivers/audio/wm8904.c:4]: Official NXP copyright header is `Copyright 2024 NXP` all

File: samples/drivers/i2s/i2s_codec/boards/mimxrt595_evk_mimxrt595s_cm33.conf [hunk 13]
```diff
@@ -0,0 +1,16 @@
+#
+# Copyright (c) 2023, NXP
+#
```
Comment (dbaluta) [inline samples/drivers/i2s/i2s_codec/boards/mimxrt595_evk_mimxrt595s_cm33.conf:3]: Fix copyright.

File: samples/drivers/i2s/i2s_codec/src/main.c [hunk 14]
```diff
@@ -0,0 +1,219 @@
+/*
+ * Copyright (c) 2024 NXP
+ *
```
Comment (dbaluta) [inline samples/drivers/i2s/i2s_codec/src/main.c:3]: Fix copyright here also. No `(c)` in the NXP recommended way of adding copyright. Just use `Copyright NXP 2024`

**Review by dbaluta — DISMISSED** [review]

File: drivers/audio/wm8904.c [hunk 15]
```diff
@@ -0,0 +1,759 @@
...
+	 * [8]   - MUTE: leaving intact
+	 * [7]   - VU: 0 (values are "cached")
+	 * [6]   - ZC: 1 (zero-crossing enabled)
+	 * [5:0] - VOL: 6-bit volume value
+	 */
+
+	const uint16_t regval = (volume | 0b001000000) & 0b011111111;
+	const uint16_t mask = 0b011111111;
+
+	switch (channel) {
```
Comment (iuliana-prodan) [inline drivers/audio/wm8904.c:227]: This switch I would write it like:
```
uint8_t reg;
switch (channel) {
case AUDIO_CHANNEL_FRONT_LEFT:
	reg = WM8904_ANALOG_OUT2_LEFT;
	break;

case AUDIO_CHANNEL_FRONT_RIGHT:
	reg = WM8904_ANALOG_OUT2_RIGHT;
	break;

case AUDIO_CHANNEL_HEADPHONE_LEFT:
	reg = WM8904_ANALOG_OUT1_LEFT;
	break;

case AUDIO_CHANNEL_HEADPHONE_RIGHT:
	reg = WM8904_ANALOG_OUT1_RIGHT;
	break;

case AUDIO_CHANNEL_ALL:
	/* Set volume values with VU = 0 */
	reg = WM8904_ANALOG_OUT1_LEFT | WM8904_ANALOG_OUT1_RIGHT | WM8904_ANALOG_OUT2_LEFT | WM8904_ANALOG_OUT2_RIGHT;
	break;
default:
	return -EINVAL;
}

wm8904_update_reg(dev, reg, mask, regval);
return 0;
```

I see you do like this, here: https://github.com/zephyrproject-rtos/zephyr/pull/75943/commits/6f41b45b2320d72424dd58fe188eeb257da57228#diff-abe7424de1cefb3ba684964873bb5830e03f03ec4dd059ea2c67c07a8102548aR376

File: drivers/audio/wm8904.c [hunk 16]
```diff
@@ -0,0 +1,759 @@
...
+	/**
+	 * WM8904_ANALOG_OUT1_LEFT, WM8904_ANALOG_OUT1_RIGHT (headphone outs),
+	 * WM8904_ANALOG_OUT2_LEFT and WM8904_ANALOG_OUT2_RIGHT (line outs):
+	 * [8]   - MUTE: mute bit
+	 * [7]   - VU: 0 (values are "cached")
+	 * [6]   - ZC: 1 (zero-crossing enabled)
+	 * [5:0] - VOL: leaving intact
+	 */
+
+	const uint16_t regval = ((mute << 8) | 0b001000000);
```
Comment (iuliana-prodan) [inline drivers/audio/wm8904.c:268]: same as above

File: drivers/audio/wm8904.c [hunk 17]
```diff
@@ -0,0 +1,759 @@
...
+
+static int wm8904_in_volume_config(const struct device *dev, audio_channel_t channel, int volume)
+{
+	/**
+	 * WM8904_ANALOG_LEFT_IN_0, WM8904_ANALOG_RIGHT_IN_0:
+	 * [7]   - MUTE: leaving intact
+	 * [4:0] - VOL: 5 bit volume value
+	 */
+
+	const uint16_t regval = volume & 0b00011111;
```
Comment (iuliana-prodan) [inline drivers/audio/wm8904.c:308]: same as above

File: drivers/audio/wm8904.c [hunk 18]
```diff
@@ -0,0 +1,759 @@
...
+
+static int wm8904_in_mute_config(const struct device *dev, audio_channel_t channel, bool mute)
+{
+	/**
+	 * WM8904_ANALOG_LEFT_IN_0, WM8904_ANALOG_RIGHT_IN_0:
+	 * [7]   - MUTE: mute bit
+	 * [4:0] - VOL: leaving intact
+	 */
+
+	const uint16_t regval = ((mute << 7) & 0b10000000);
```
Comment (iuliana-prodan) [inline drivers/audio/wm8904.c:338]: same here

File: drivers/audio/wm8904.c [hunk 19]
```diff
@@ -0,0 +1,759 @@
...
+
+	default:
+		break;
+	}
+
+	return 0;
+}
+
+static void wm8904_start_output(const struct device *dev)
+{
```
Comment (iuliana-prodan) [inline drivers/audio/wm8904.c:623]: Why are these 2 functions empty? 
No logging info, nothing?

File: drivers/audio/wm8904.c [hunk 20]
```diff
@@ -0,0 +1,759 @@
...
+	return -EINVAL;
+}
+
+static int wm8904_apply_properties(const struct device *dev)
+{
+	/**
+	 * Set VU = 1 for all output channels, VU takes effect for the whole
+	 * channel pair.
+	 */
+	wm8904_update_reg(dev, WM8904_ANALOG_OUT1_LEFT, 0b010000000, 0b010000000);
```
Comment (iuliana-prodan) [inline drivers/audio/wm8904.c:656]: Use macros

File: samples/drivers/i2s/i2s_codec/prj.conf [hunk 21]
```diff
@@ -0,0 +1,4 @@
+CONFIG_I2S=y
+#CONFIG_LOG=y
```
Comment (iuliana-prodan) [inline samples/drivers/i2s/i2s_codec/prj.conf:2]: If you don't use it, delete it

**Review by iuliana-prodan — CHANGES_REQUESTED** [review]

File: drivers/audio/wm8904.c [hunk 22]
```diff
@@ -0,0 +1,759 @@
...
+		LOG_ERR("Word size %d bits not supported; falling back to 16 bits",
+			cfg->i2s.word_size);
+		word_size = 0;
+		break;
+	}
+	/* Set bit resolution */
+	wm8904_update_reg(dev, WM8904_AUDIO_IF_1, (0x000CU), ((uint16_t)(word_size) << 2U));
+	{
+		uint16_t reg_val = 0;
+
```
[the remark this answers is not in the document]
Comment (VitekST) [reply drivers/audio/wm8904.c:203]: It's not, it's just an artifact that was left unattended.

File: drivers/audio/wm8904.c (hunk 15 above)
Comment (VitekST) [reply drivers/audio/wm8904.c:227]: That can't be done, as the `reg` variable holds an ordinal register index, it isn't a flag-style variable with disjunctive bits.

File: drivers/audio/wm8904.c (hunk 19 above)
Comment (VitekST) [reply drivers/audio/wm8904.c:623]: @iuliana-prodan These functions correspond to the `audio_codec_start_output` and `audio_codec_stop_output` syscalls in `include/zephyr/audio/codec.h`. These syscall functions don't permit NULL values in the API structure, thus these functions are mandatory to implement. My goal was to change the `codec` API as little as possible, so that's why I left it as it is.

File: drivers/audio/wm8904.c [hunk 23]
```diff
@@ -0,0 +1,759 @@
...
+	/**
+	 * WM8904_ANALOG_OUT1_LEFT, WM8904_ANALOG_OUT1_RIGHT (headphone outs),
+	 * WM8904_ANALOG_OUT2_LEFT and WM8904_ANALOG_OUT2_RIGHT (line outs):
+	 * [8]   - MUTE: leaving intact
+	 * [7]   - VU: 0 (values are "cached")
+	 * [6]   - ZC: 1 (zero-crossing enabled)
+	 * [5:0] - VOL: 6-bit volume value
+	 */
+
+	const uint16_t regval = (volume | 0b001000000) & 0b011111111;
```
[the remark this answers is not in the document]
Comment (VitekST) [reply drivers/audio/wm8904.c:218]: @iuliana-prodan Do they? I personally don't see how that would be beneficial. The mask and content values differ practically for each register update operation, they are based directly on the codec's datasheet and, at least the ones I've added, are annotated. 

Does your comment apply also for other invocations of `wm8904_update_reg`, which here originally authored by Hake?

File: drivers/audio/wm8904.c (hunk 15 above)
Comment (iuliana-prodan) [reply drivers/audio/wm8904.c:215]: ok, got it.

File: drivers/audio/wm8904.c (hunk 19 above)
Comment (iuliana-prodan) [reply drivers/audio/wm8904.c:583]: ok.

**Review by iuliana-prodan — DISMISSED** [review]

**Review by dbaluta — DISMISSED** [review]

**Review by dbaluta — DISMISSED** [review]

**Comment by dbaluta** [comment]
@VitekST  looks good to merge! We tested it on imxrt 685 and works fine.

**Review by hakehuang — DISMISSED** [review]

**Review by iuliana-prodan — DISMISSED** [review]

**Comment by VitekST** [comment]
@anangl Is everything up to your standards?

File: drivers/audio/wm8904.c [hunk 24]
```diff
@@ -0,0 +1,728 @@
...
+{
+	LOG_DBG("Initializing codec.");
+
+	const struct wm8904_driver_config *const dev_cfg = DEV_CFG(dev);
+
+	int error = pinctrl_apply_state(dev_cfg->pincfg, PINCTRL_STATE_DEFAULT);
+
+	if (error) {
+		LOG_ERR("pinctrl_apply_state fail: %d", error);
+		return error;
```
Comment (anangl) [inline drivers/audio/wm8904.c:51]: This makes the "default" pinctrl state required but "pinctrl-0" and "pinctrl-names" properties are not marked so in the binding. But is actually the external codec driver the one that should take care of configuring SoC pins? This looks a bit strange. @gmarull

File: drivers/audio/wm8904.c [hunk 25]
```diff
@@ -0,0 +1,728 @@
...
+	}
+
+	error = i2c_recover_bus(dev_cfg->i2c.bus);
```
Comment (anangl) [inline drivers/audio/wm8904.c:54]: Why does this driver always need to recover the I2C bus?

File: drivers/audio/wm8904.c [hunk 26]
```diff
@@ -0,0 +1,728 @@
...
+		return -EINVAL;
+	}
+}
+
+static int wm8904_out_mute_config(const struct device *dev, audio_channel_t channel, bool mute)
+{
+	const uint16_t val = WM8904_REGVAL_OUT_VOL(mute, 0, 0, 0);
+	const uint16_t mask = WM8904_REGMASK_OUT_MUTE;
+
+	switch (channel) {
```
Comment (anangl) [inline drivers/audio/wm8904.c:249]: This `switch' is the same as in `wm8904_out_mute_config()`. Wouldn't it make sense to use a common part of code for both functions?

File: drivers/audio/wm8904.c [hunk 27]
```diff
@@ -0,0 +1,728 @@
...
+		return -EINVAL;
+	}
+}
+
+static int wm8904_in_mute_config(const struct device *dev, audio_channel_t channel, bool mute)
+{
+	const uint16_t val = WM8904_REGVAL_IN_VOL(mute, 0);
+	const uint16_t mask = WM8904_REGMASK_IN_MUTE;
+
+	switch (channel) {
```
Comment (anangl) [inline drivers/audio/wm8904.c:307]: Similarly here. It is the same as in `wm8904_in_mute_config()`.

File: drivers/audio/wm8904.c [hunk 28]
```diff
@@ -0,0 +1,728 @@
...
+	}
+}
+
+static void wm8904_update_reg(const struct device *dev, uint8_t reg, uint16_t mask, uint16_t val)
+{
+	uint16_t reg_val = 0;
+	uint16_t new_value = 0;
+
+	if (reg == 0x19) {
+		LOG_INF("try write mask %#x val %#x", mask, val);
```
Comment (anangl) [inline drivers/audio/wm8904.c:673]: These `LOG_INF()` calls in this function will probably generate a lot of logging messages. Wouldn't `LOG_DBG()` be a better choice?

File: boards/nxp/mimxrt595_evk/board.c [hunk 29]
```diff
@@ -279,34 +279,32 @@ static int mimxrt595_evk_init(void)
 	pm_policy_state_lock_get(PM_STATE_SUSPEND_TO_IDLE, PM_ALL_SUBSTATES);
 
 #ifdef CONFIG_I2S
-
+#ifdef CONFIG_I2S_TEST_SEPARATE_DEVICES
```
Comment (anangl) [inline boards/nxp/mimxrt595_evk/board.c:282]: This Kconfig option is only used by I2S tests. It's quite weird that it is handled in some board initialization code.

File: drivers/audio/wm8904.c [hunk 30]
```diff
@@ -0,0 +1,728 @@
...
+{
+	wm8904_route_input(dev, AUDIO_CHANNEL_FRONT_LEFT, 2);
+	wm8904_route_input(dev, AUDIO_CHANNEL_FRONT_RIGHT, 2);
+
+	wm8904_in_volume_config(dev, AUDIO_CHANNEL_ALL, WM8904_INPUT_VOLUME_DEFAULT);
+	wm8904_in_mute_config(dev, AUDIO_CHANNEL_ALL, false);
+}
+
+static const struct audio_codec_api wm8904_driver_api = {.configure = wm8904_configure,
+							 .start_output = wm8904_start_output,
```
[the remark this answers is not in the document]
Comment (VitekST) [reply drivers/audio/wm8904.c:709]: That is a product of `clang-format`. Am I not supposed to use it at all for new files, despite recommendations in contributor's guidelines?

File: drivers/audio/wm8904.c (hunk 25 above)
Comment (VitekST) [reply drivers/audio/wm8904.c:54]: @anangl Looks like a requirement when trying to use RT595/RT685's I3C peripheral as an I2C master.

File: boards/nxp/mimxrt595_evk/board.c (hunk 29 above)
Comment (VitekST) [reply boards/nxp/mimxrt595_evk/board.c:282]: @anangl It's true that this property could be named differently (as I didn't author that name), I believe it has its place in either soc.c or board.c. This property implies how the signals between Flexcomm peripherals in an I2S mode should be shared. Duplex audio on the RTxxx devices with a single codec is handled by configuring one peripheral as an I2S master and the other as a slave. The slave peripheral then has the LRCK (WS) and SCK signals as inputs. It may be desirable not to set up signal sharing, for example when these devices are used with separate ADCs and DACs.

File: drivers/audio/wm8904.c (hunk 27 above)
Comment (VitekST) [reply drivers/audio/wm8904.c:307]: @anangl Personally I think that the code can be followed a bit more easily than if one strives to avoid even two-fold repetitions at all cost, but if you insist...

**Review by hakehuang — DISMISSED** [review]

File: drivers/audio/wm8904.c (hunk 24 above)
Comment (VitekST) [reply drivers/audio/wm8904.c]: @anangl Looks like another artifact from the previous author. I am also inclined to keep pinctrl configuration in the realm of drivers driving peripherals that expose used signals instead of the consumers and personally I don't see any reason not to configure the MCLK signal even though it wasn't done so before. Thus removed.

**Review by dbaluta — APPROVED** [review]

**Review by hakehuang — APPROVED** [review]

**Review by butok — APPROVED** [review]

**Review by iuliana-prodan — APPROVED** [review]
